### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by ykshatroff**
This adds authentication with ID Card, MobileID and SmartID. Also, this fixes some issues in signing, updates SmartID endpoints as per latest changes (November 2020), and updates documentation.

Other changes besides mentioned above:
* Added pylint.
* Fixed the signing `SESSION_VALIDITY_TIMEOUT`.
* Fixed a problem with errors not being handled properly in the SessionViewMixin.
* Moved `PredictableDict` to the core `types`.
* Moved common stuff to `mixins`.

### CODE REVIEW

File: esteid/idcard/README.md
Comment (Jyrno42) [inline esteid/idcard/README.md:98]: > The multi-domain approach (i.e. each new authentication happens on a different domain) didn't help.

Just curious, has something changed or why did it not help? Previously when the subdomain was just a random hash then browsers just asked for the pin again (when the card in the reader changed).

Thinking that maybe the behavior changed with the fix for the recent idcard security issue.

File: esteid/idcard/README.md
Comment (Jyrno42) [inline esteid/idcard/README.md:80]: Not really sure I would have added iframes to the mix tbh. 

In etk I opened a new browser window which notified its parent about things with the same postMessage trick (window.open in original page and window.opener to use postMessage).

This way X-Frame-Options can be kept as `DENY` and CSP should not matter.

File: esteid/mobileid/types.py
Comment (Jyrno42) [inline esteid/mobileid/types.py:65]: I don't think this should be validated by esteid as this regex based approach might be too fragile. I would remove the phone number validation from this method and would just expect users to pass in validated values.

> I hope the MID/SmartID rest api will raise an error in case the provided data is not valid.

**Review by Jyrno42 — CHANGES_REQUESTED** [review]

File: esteid/idcard/README.md
Comment (Jyrno42) [reply esteid/idcard/README.md:80]: What are your thoughts on the method without iframes?

File: esteid/idcard/README.md
Comment (Jyrno42) [reply esteid/idcard/README.md:98]: https://www.id.ee/artikkel/ria-koos-partneritega-parandas-id-kaardi-brauserilaienduses-kriitilise-vea/

**Comment by Jyrno42** [comment]
Great work. 

Maybe it makes sense to add a section about why (with nginx) the certificate is validated with OCSP not trough revocation lists. Basically just linking to the following things:

- https://trac.nginx.org/nginx/ticket/1094
- http://mailman.nginx.org/pipermail/nginx-devel/2017-March/009609.html

File: esteid/idcard/README.md
Comment (ykshatroff) [reply esteid/idcard/README.md:80]: I feel it's not a very fine solution.
Window.open seems to be blocked by default (in Firefox, at least, it shows this yellow bar 'prevented a popup') and it's definitely not the best UX even if it weren't blocked, because it may be affected by window focusing preferences and there would be either tabs jumping or windows in the taskbar.
X-Frame-Options can still be kept as DENY for the main site, and the `auth` site would only need to enable one specific location, with everything else behind 404.
Also as per this TR https://www.w3.org/TR/CSP2/#frame-ancestors-and-frame-options , X-Frame-Options is called obsolete :)

File: esteid/mobileid/types.py
Comment (ykshatroff) [reply esteid/mobileid/types.py:65]: I don't think the phone numbering scheme for Estonia and Lithuania is going to change soon :)
I could make the regexp accept broader values (like 7-10 digits maybe), but I'd leave at least some sanity check in place. Otherwise this could result in us being a mediator in DDoS attacks :)

File: esteid/idcard/README.md
Comment (ykshatroff) [reply esteid/idcard/README.md:98]: Can't say if the news has anything to do with that (I rather felt it was about a possibility to pass on the info received from the ID card by e.g. nginx, to use it to access a 3rd service e.g. eesti.ee) but I started with trying the subdomain model (I remember we discussed it in slack) and whenever I got authenticated on one subdomain, I was also authenticated on others. I was using firefox because in Chrome this plugin just doesn't see the ID card. Since it's some opaque mechanism I couldn't debug, I can't really suggest any solution other than just get along with it for now. Maybe it could be found out during "production testing" ...
I would say this whole thing is a bit unreliable, as sometimes it asks me to select certificate, sometimes it doesn't, and even if I take the ID card out of the reader it kind of doesn't notice it and I keep being authenticated.

File: esteid/idcard/README.md
Comment (Jyrno42) [reply esteid/idcard/README.md:80]: window.open is not blocked when called from a user event handler (like onclick)

Comment (Jyrno42) [reply esteid/idcard/README.md:80]: New tabs/windows are also used for Oauth flows so I don't really agree with your point about UX.

File: esteid/mobileid/types.py
Comment (Jyrno42) [reply esteid/mobileid/types.py:65]: I would still leave that to be something to deal with in the userland. Phone number validation is not really the scope of this project.

File: esteid/idcard/README.md
Comment (ykshatroff) [reply esteid/idcard/README.md:80]: > window.open is not blocked when called from a user event handler (like onclick)

forgot about it, thanks :)

For OAuth, when you need to enter data in a form, then it makes sense to open up a window, But a window that bears 0 significance to the user seems to me to be useless and not good. I also haven't seen any other services (LHV, Swedbank or Estonian government services) which would use a window.

And all in all, I don't really see any drawbacks to the iframe approach.

File: esteid/mobileid/types.py
Comment (ykshatroff) [reply esteid/mobileid/types.py:65]: The problem is that userland can also make mistakes and forget about sanity checks. If we don't do a sanity check, we're still open to trouble. Also as we do ID code validation, it would be logical to do at least _some_ phone validation. Phones will always consist of `+`, country code , and several digits, the question is how many digits :) If MID starts supporting other countries, I guess it could be adjusted.
This is by no means a complete validation, I'd prefer to call it sanity check.

File: esteid/idcard/README.md
Comment (Jyrno42) [reply esteid/idcard/README.md:80]: the banks navigate you trough the id auth domain/location and then redirect you back. all in the same window. 

but this is nitpicking anyway and not too relevant

File: esteid/mobileid/types.py
Comment (Jyrno42) [reply esteid/mobileid/types.py:65]: What trouble are you worried about exactly? This is just an api call which has validation on their side anyway.

In the end I would also be happy if it would be possible to easily override this validation in the userland (currently I can't even monkeypatch this as all validation occurs in the same method).

Comment (ykshatroff) [reply esteid/mobileid/types.py:65]: I'm worried that potential requests with invalid data can lead to a) performance degradation of the app itself because MID is not the fastest responding service b) problems with MID as provider that we (i.e. the client) are not performing the necessary validation.

I was just about to suggest that I could make it optional, adding a feature switch. This would help obviate potential issues with it.

Comment (ykshatroff) [reply esteid/mobileid/types.py:65]: Or just make this `PHONE_NUMBER_REGEXP` configurable, what do you think?

Comment (ykshatroff) [reply esteid/mobileid/types.py:65]: Created a PR #53 with this check made optional and configurable.

File: esteid/idcard/README.md
Comment (ykshatroff) [reply esteid/idcard/README.md:98]: Actually you're right, I gave it another round of tests and it seems to be working with different subdomains now. Might have been some other problem such as misconfiguration... Sorry for that.

I'm happy to make the necessary changes, but _for local testing_ there is a conundrum - if every auth frame is opened with a new domain, then before it is usable, one needs to add a security exception for that new domain (due to self-signed cert), or else the frame is blocked. So had to add a bit of guidance through this, and - another fun thing - this doesn't work in Firefox because it asks the ID card PIN *before* asking for a security exception, and this means that the frame itself will not ask the ID card credentials (at least for the period of caching). Luckily, for some reason the plugin started working in Chrome (maybe a version update helped) so it feels this concept is usable.

Again, my apologies for not being thorough from the start.

**Review by Jyrno42 — APPROVED** [review]

**Pull request merged**
